lib/output/writer/kinesis_firehose.go

```diff
@@ -0,0 +1,260 @@
+func NewKinesisFirehoseConfig() KinesisFirehoseConfig {
+	rConf := retries.NewConfig()
+	rConf.Backoff.InitialInterval = "1s"
+	rConf.Backoff.MaxInterval = "5s"
+	rConf.Backoff.MaxElapsedTime = "30s"
+	return KinesisFirehoseConfig{
+		sessionConfig: sessionConfig{
+			Config: sess.NewConfig(),
+		},
+		Stream: "",
+		Config: rConf,
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// KinesisFirehose is a benthos writer.Type implementation that writes messages
+// to an Amazon Kinesis Firehose destination.
+type KinesisFirehose struct {
+	conf KinesisFirehoseConfig
+
+	session  *session.Session
+	firehose firehoseiface.FirehoseAPI
+
+	backoff    backoff.BackOff
+	endpoint   *string
+	streamName *string
+
+	log   log.Modular
+	stats metrics.Type
+
+	mThrottled       metrics.StatCounter
+	mThrottledF      metrics.StatCounter
+	mPartsThrottled  metrics.StatCounter
+	mPartsThrottledF metrics.StatCounter
+}
+
+// NewKinesisFirehose creates a new Amazon Kinesis Firehose writer.Type.
+func NewKinesisFirehose(
+	conf KinesisFirehoseConfig,
+	log log.Modular,
+	stats metrics.Type,
+) (*KinesisFirehose, error) {
+	k := KinesisFirehose{
+		conf:            conf,
+		log:             log,
+		stats:           stats,
+		mPartsThrottled: stats.GetCounter("parts.send.throttled"),
+		mThrottled:      stats.GetCounter("send.throttled"),
+		streamName:      aws.String(conf.Stream),
+	}
+
+	var err error
+	if k.backoff, err = conf.Config.Get(); err != nil {
+		return nil, err
+	}
+	return &k, nil
+}
+
+//------------------------------------------------------------------------------
+
+// toRecords converts an individual benthos message into a slice of Kinesis Firehose
+// batch put entries by promoting each message part into a single part message
+// and passing each new message through the partition and hash key interpolation
+// process, allowing the user to define the partition and hash key per message
+// part.
+func (a *KinesisFirehose) toRecords(msg types.Message) ([]*firehose.Record, error) {
+	entries := make([]*firehose.Record, msg.Len())
+
+	err := msg.Iter(func(i int, p types.Part) error {
+		message.Lock(msg, i)
+
+		entry := firehose.Record{
+			Data: p.Get(),
+		}
+
+		if len(entry.Data) > mebibyte {
+			a.log.Errorf("part %d exceeds the maximum Kinesis Firehose payload limit of 1 MiB\n", i)
+			return types.ErrMessageTooLarge
+		}
+
+		entries[i] = &entry
+		return nil
+	})
+
+	return entries, err
+}
+
+//------------------------------------------------------------------------------
+
+// Connect creates a new Kinesis Firehose client and ensures that the target Kinesis
+// Firehose delivery stream.
+func (a *KinesisFirehose) Connect() error {
+	if a.session != nil {
+		return nil
+	}
+
+	sess, err := a.conf.GetSession()
+	if err != nil {
+		return err
+	}
+
+	a.session = sess
+	a.firehose = firehose.New(sess)
+
+	if _, err := a.firehose.DescribeDeliveryStream(&firehose.DescribeDeliveryStreamInput{
```

Comment:
It might be worth adding a more thorough check to ensure that the particular Kinesis Firehose is not just deployed, but also is deployed in a way that we can actually send messages directly to it (instead of being attached to service directly like Kinesis etc).  This would just check the response for `DeliveryStreamDescription.DeliveryStreamType == "DirectPut"`.

I will hold off on this in case there is additional feedback.